forge/csrc/passes/split_graph.cpp

```diff
@@ -333,26 +332,147 @@ std::unique_ptr<Graph> extract_backward_graph(
     return bwd_graph;
 }
 
+std::unique_ptr<Graph> extract_optimizer_graph(
+    const Graph *graph, const Graph *fwd_graph, const std::vector<graphlib::Node *> &topo)
+{
+    auto opt_graph = std::make_unique<Graph>(tt::graphlib::IRLevel::IR_TT_FORGE, "optimizer");
+    opt_graph->set_training(graph->training());
+
+    // Adding all trainable params (which require gradients) from the forward graph as inputs to the optimizer graph.
+    std::vector<graphlib::NodeId> opt_module_outputs;
+    for (auto param : graph->get_parameter_nodes())
+    {
+        if (param->as<graphlib::InputNode>()->requires_grad())
+        {
+            TT_ASSERT(
+                has_users_in_opt(graph, param),
+                "Expected parameter node {} to have users in the optimizer graph",
+                param->name());
+            log_debug("Adding parameter node {} as input to opt graph", param->name());
+            clone_and_add(param, opt_graph.get());
+
+            // For each trainable parameter, we need to create the output node for the updated parameter (aliased to the
+            // input node). E.g. If the parameter is `w`, then the output node will be `w_updated`: `w_updated = w - lr
+            // * gradient`.
+            //
+            // The runtime will look for aliased outputs and will make sure that the appropriate tensors are updated.
+            // See `OutputNode::is_aliased_tensor()` for more details.
+            auto grad_output_node = graphlib::create_node<graphlib::OutputNode>(param->name() + "_updated");
+            grad_output_node->set_output_type(graphlib::OutputType::Internal);
+            grad_output_node->set_shape(param->shape());
+            grad_output_node->set_output_df(param->output_df());
+            grad_output_node->set_epoch_type(graphlib::NodeEpochType::Optimizer);
+            grad_output_node->set_alias(param->as<graphlib::InputNode>());
+
+            opt_graph->add_node(std::move(grad_output_node), 0 /*subgraph_id=*/);
+            opt_module_outputs.push_back(opt_graph->get_node_by_name(param->name() + "_updated")->id());
+        }
+    }
+
+    std::vector<graphlib::NodeId> opt_module_inputs;
+
+    // All gradients computed in the backward graphs should be inputs to the optimizer graph - add them.
```

Comment:
All *parameter* gradients computed in the bwd are inputs to opt, activation gradients aren't - this might well be the case here but it's not obvious to me.